sound/soc/sof/topology.c

```diff
@@ -3274,71 +3292,88 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 	 * DAI links are expected to have at least 1 hw_config.
 	 * But some older topologies might have no hw_config for HDA dai links.
 	 */
-	num_hw_configs = le32_to_cpu(cfg->num_hw_configs);
-	if (!num_hw_configs) {
-		if (config.type != SOF_DAI_INTEL_HDA) {
+	hw_config = cfg->hw_config;
+	num_conf = le32_to_cpu(cfg->num_hw_configs);
+	if (!num_conf) {
+		if (common_config.type != SOF_DAI_INTEL_HDA) {
 			dev_err(scomp->dev, "error: unexpected DAI config count %d!\n",
 				le32_to_cpu(cfg->num_hw_configs));
 			return -EINVAL;
 		}
+		num_conf = 1;
 	} else {
 		dev_dbg(scomp->dev, "tplg: %d hw_configs found, default id: %d!\n",
 			cfg->num_hw_configs, le32_to_cpu(cfg->default_hw_config_id));
 
-		for (i = 0; i < num_hw_configs; i++) {
-			if (cfg->hw_config[i].id == cfg->default_hw_config_id)
+		for (curr_conf = 0; curr_conf < num_conf; curr_conf++) {
+			if (hw_config[curr_conf].id == cfg->default_hw_config_id)
 				break;
 		}
 
-		if (i == num_hw_configs) {
+		if (curr_conf == num_conf) {
 			dev_err(scomp->dev, "error: default hw_config id: %d not found!\n",
 				le32_to_cpu(cfg->default_hw_config_id));
 			return -EINVAL;
 		}
+
+		/* Currently only SSP supports multiple configs, for other DAI's only default */
+		/* config is parsed and saved */
+		if (common_config.type != SOF_DAI_INTEL_SSP)
+			num_conf = 1;
 	}
 
-	/* configure dai IPC message */
-	hw_config = &cfg->hw_config[i];
+	/* Reserve memory for all hw configs, eventually freed by widget */
+	config = kcalloc(num_conf, sizeof(*config), GFP_KERNEL);
+	if (!config)
+		return -ENOMEM;
 
-	config.hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
-	config.format = le32_to_cpu(hw_config->fmt);
+	/* Copy common data to all config ipc structs */
+	if (num_conf > 1) {
+		for (i = 0; i < num_conf; i++) {
+			config[i].hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
+			config[i].format = hw_config[i].fmt;
+			config[i].type = common_config.type;
+			config[i].dai_index = common_config.dai_index;
+		}
+	} else {
+		config[0].hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
+		config[0].format = hw_config[curr_conf].fmt;
+		config[0].type = common_config.type;
+		config[0].dai_index = common_config.dai_index;
+		hw_config += curr_conf;
```

Comment:
Is this else part really necessary? If num_conf is 1, curr_conf will be 0 no?

Reply:
Ack, I think you can just do the for loop.

Reply:
not necessary if you have defined multiple configs in your topology and you default conf doesn't point to the first one. This is unlikely but possible. I set num_conf forcibly to 1 if this is not SSP dai. That was the previous functionality : discard other possible defined configs than the default conf.

Reply:
oh that is confusing. so you're saying that there're multiple configs but you force it down to 1. Lets not do that. I think the better way is to just leave the multiple configs and set the default config.

Reply:
ok